docs/rfcs/022-pageserver-delete-from-s3.md

```diff
@@ -0,0 +1,290 @@
+    participant CP as Control Plane
+    participant PS as Pageserver
+    participant S3
+
+    CP->>PS: Delete tenant
+    PS->>S3: Create deleted mark file at <br> /tenant/meta/deleted
+    PS->>PS: Create deleted mark file locally
+    PS->>CP: Accepted
+    PS->>PS: delete local files other than deleted mark
+    loop Delete layers for each timeline
+        PS->>S3: delete(..)
+        CP->>PS: Finished?
+        PS->>CP: False
+    end
+    PS->>S3: Delete mark file
+    PS->>PS: Delete local mark file
+
+    loop Poll for status
+        CP->>PS: Finished?
+        PS->>CP: True or False
+    end
+```
+
+Why two mark files?
+Remote one is needed for cases when pageserver is lost during deletion so other pageserver can learn the deletion from s3 during attach.
+
+Local mark file is needed for case when pageserver crashes after remote one is deleted b
+TODO describe mark files
+Why local mark file is needed? If we dont have one, we have two choices, delete local data before deleting the remote part or do that after.
+
+If we delete local data before remote then during restart pageserver wont pick up remote tenant at all because nothing is available locally (pageserver looks for remote conuterparts of locally available tenants).
+
+If we delete local data after remote then at the end of the sequence when remote mark file is deleted if pageserver restart happens then the state is the same to situation when pageserver just missing data on remote without knowing the fact that this data is deleted, in this case the current behavior is upload everything has local data and remote
+
+Thus we need local record of tenant being deleted as well.
+
+##### Handle pageserver crashes
+
+Lets explore sequences with various crash points.
+
+Pageserver crashes before `deleted` mark file is persisted in s3:
+
+```mermaid
+sequenceDiagram
+    autonumber
+    participant CP as Control Plane
+    participant PS as Pageserver
+    participant S3
+
+    CP->>PS: Delete tenant
+    note over PS: Crash point 1.
+    CP->>PS: Retry delete request
+
+    PS->>S3: Create deleted mark file at <br> /tenant/meta/deleted
+    PS->>PS: Create deleted mark file locally
+
+    PS->>CP: Accepted
+
+    PS->>PS: delete local files other than deleted mark
+
+    loop Delete layers for each timeline
+        PS->>S3: delete(..)
+        CP->>PS: Finished?
+        PS->>CP: False
+    end
+    PS->>S3: Delete mark file
+    PS->>PS: Delete local mark file
+
+    CP->>PS: Finished?
+    PS->>CP: True
+```
+
+Pageserver crashed when deleted mark was about to be persisted in s3, before Control Plane gets a response:
+
+```mermaid
+sequenceDiagram
+    autonumber
+    participant CP as Control Plane
+    participant PS as Pageserver
+    participant S3
+
+    CP->>PS: Delete tenant
+    PS->>S3: Create deleted mark file at <br> /tenant/meta/deleted
+
+    note over PS: Crash point 2.
+    note over PS: During startup we reconcile <br> with remote and see <br> whether the mark exists
```

Comment:
Reconcile will just need to continue from the "create mark locally" phase right, or is the local mark file necessary (assuming we will always check for s3 mark file before-hand)?

Reply:
Looking now, probably you omitted the local mark file because it's difficult to keep the two in sync or unnecessary detail?

Reply:
> Reconcile will just need to continue from the "create mark locally" phase right

Correct

> Looking now, probably you omitted the local mark file because it's difficult to keep the two seq diagrams in sync or unnecessary detail?

You're right, I added local mark everywhere after remote one, so here it remained to be just "mark". I'll clarify the comment.

The idea is to create remote mark first, so in both cases (restart and attach we can figure it out)

Reply:
Ok, makes sense. SomeoneToIgnore already created a thread about the essence of mark files so resolving this in favor of those:
- https://github.com/neondatabase/neon/pull/3792#discussion_r1133861913
- ... thought there was another one, but must've misunderstood.